core/wasm-interface/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+}
+
+impl Signature {
+	/// Create a new instance of `Signature`.
+	pub fn new<T: Into<Cow<'static, [ValueType]>>>(args: T, return_value: Option<ValueType>) -> Self {
+		Self {
+			args: args.into(),
+			return_value,
+		}
+	}
+
+	/// Create a new instance of `Signature` with the given `args` and without any return value.
+	pub fn new_with_args<T: Into<Cow<'static, [ValueType]>>>(args: T) -> Self {
+		Self {
+			args: args.into(),
+			return_value: None,
+		}
+	}
+
+}
+
+/// A reference to a host function.
+#[derive(Eq, PartialEq, Debug)]
+pub struct FunctionRef {
+	/// The signature of the function.
+	pub signature: Signature,
+	/// The index of the function at the host.
+	pub index: usize,
+}
+
+impl FunctionRef {
+	/// Create a new instance of `FunctionRef`.
+	pub fn new(signature: Signature, index: usize) -> Self {
+		Self {
+			signature,
+			index,
+		}
+	}
+
+	/// Add the given offset to the internal index.
+	pub fn offset_index(&mut self, offset: usize) {
+		self.index += offset;
+	}
+}
+
+/// Context used by `Externals` to interact with the allocator and the memory of the wasm instance.
+pub trait ExternalsContext {
+	/// Read memory from `address` into a vector.
+	fn read_memory(&self, address: Pointer<u8>, size: WordSize) -> Result<Vec<u8>, String> {
+		let mut vec = Vec::with_capacity(size as usize);
+		unsafe { vec.set_len(size as usize); }
+		self.read_memory_into(address, &mut vec)?;
+		Ok(vec)
+	}
+	/// Read memory into the given `dest` buffer from `address`.
+	fn read_memory_into(&self, address: Pointer<u8>, dest: &mut [u8]) -> Result<(), String>;
+	/// Write the given data at `address` into the memory.
+	fn write_memory(&mut self, address: Pointer<u8>, data: &[u8]) -> Result<(), String>;
+	/// Allocate a memory instance of `size` bytes.
+	fn allocate_memory(&mut self, size: WordSize) -> Result<Pointer<u8>, String>;
+	/// Deallocate a given memory instance.
+	fn deallocate_memory(&mut self, ptr: Pointer<u8>) -> Result<(), String>;
+}
+
+/// Something that provides implementations for host functions.
+pub trait Externals {
+	/// Try to resolve the function with the given name and signature.
+	fn resolve_function(name: &str, signature: &Signature) -> Result<Option<FunctionRef>, String>;
+
+	/// Returns the number of host functions this type provides.
+	fn function_count() -> usize;
+
+	/// Execute the function at the given index.
+	///
+	/// - `index` - Is equal to the index given to `FunctionRef`.
+	/// - `args` - The arguments given to the function.
+	/// - `context` - Provides access to the allocator and memory of the memory wasm instance.
+	fn execute_function<A: Iterator<Item=Value>>(
+		index: usize,
+		args: A,
+		context: &mut dyn ExternalsContext,
+	) -> Result<Option<Value>, String>;
+}
+
+/// Something that provides implementations for inherent host functions.
+///
+/// Will be used interanlly by Substrate to provide functions like `malloc`, `free` etc.
+pub trait InherentExternals {
```

Comment:
I don't like the division between `Externals` and `InherentExternals`.

My preference would be that there be two (or more) implementations of the same interface, `Externals`, and `FunctionExecutor` keeps a `Vec<Box<dyn Externals>>` of externals. Then the `ModuleImportResolver` impl of `FunctionExecutor` would scan through all of them for the function it wants.

Reply:
If we move more stuff into the externals, we could probably remove the `InherentExternals` in the future, but I would really like to leave this for the future :) 

The general design is already suited for multiple externals, we will just add a generic Parameter `E: Externals` that can hold a tuple of externals. When you then call one function of this generic parameter, it will automatically cascade down to all the tuple elements. (Comparable to `OnInitialize` in the runtime)

Reply:
What else needs to be moved into externals? The `HostFunctions` trait is nearly identical to the `InherentHostFunctions` trait, except that one takes a context parameter and in the other one, the context is `self`. (And the passed-in context turns out to be `self` anyway`).

I really don't understand the reason for two separate traits and code paths.

Reply:
As written in the comments to the pr, we/I can remove this in a future pr. I just don't wanted to touch the `Externalities` trait in this pr. When we move the `sandbox` stuff into the `Externalities`, we can remove the `InherentHostFunctions` trait.